R/tspa_plot.R

```diff
@@ -0,0 +1,149 @@
+#' data = fs_dat,
+#' se = list(ind60 = 0.1213615, dem60 = 0.6756472))
+#' tspa_plot(tspa_fit)
+
+tspa_plot <- function (tspa_fit,
+                       ask = FALSE,
+                       title = NULL,
+                       label_x = NULL,
+                       label_y = NULL,
+                       abbreviation = TRUE,
+                       ...) {
+
+  fit_data <- parameterestimates(tspa_fit)
+  latent_scores <- lavInspect(tspa_fit, what = "data")
+
+  if (is.list(latent_scores)) {
+    latent_names <- colnames(latent_scores[[1]])
+    df_latent_scores <- lapply(latent_scores, data.frame)
+    ifelse(abbreviation == TRUE,
+           g_names <- abbreviate(names(df_latent_scores)),
+           g_names <- names(df_latent_scores))
+    latent_dv <- c(t(na.omit(fit_data[1:(nrow(fit_data)/length(latent_scores)), ][which(fit_data$op == "~"),]["lhs"])))
+    latent_iv <- c(t(na.omit(fit_data[1:(nrow(fit_data)/length(latent_scores)), ][which(fit_data$op == "~"),]["rhs"])))
+    latent_model <- list()
+
+    for (g in seq(length(df_latent_scores))) {
+      for (i in seq(length(latent_dv))) {
+        latent_model[[g]] <- lm(as.numeric(t(df_latent_scores[[g]][paste0("fs_", latent_dv[i])])) ~
+                             as.numeric(t(df_latent_scores[[g]][paste0("fs_", latent_iv[i])])),
+                           data = df_latent_scores[[g]])
+
+        if (ask == TRUE) {
+          invisible(readline(prompt = "Hit <Return> to see next plot: "))
+        }
+
+        plot(latent_scores[[g]][ ,paste0("fs_", latent_iv[i])],
+             latent_scores[[g]][ ,paste0("fs_", latent_dv[i])],
+             ylab = ifelse(is.null(label_y), paste0("fs_", latent_dv[i]),
+                           ifelse(length(label_y) > 1, label_y[i], label_y)),
+             xlab = ifelse(is.null(label_x), paste0("fs_", latent_iv[i]),
+                           ifelse(length(label_x) > 1, label_x[i], label_x)),
+             main = ifelse(is.null(title),
+                           paste0("Scatterplot", " (Group ", g, ": ", g_names[g], ")"),
+                           ifelse(length(title) > 1, title[i], title)),
+             pch = 16,
+             ...)
+        abline(latent_model[[g]])
+
+        if (ask == TRUE) {
+          invisible(readline(prompt = "Hit <Return> to see next plot: "))
+        }
+        df_latent_scores[[g]]$residuals <- latent_model[[g]]$residuals
+        plot(latent_scores[[g]][ ,paste0("fs_", latent_iv[i])],
+             df_latent_scores[[g]]$residuals,
+             ylab = "Residuals",
+             xlab = "Fitted values",
+             main = paste0("Residual Plot", " (Group ", g, ": ", g_names[g], ")"),
+             pch = 18,
+             ...)
+        abline(0, 0,
+               lty = 3)
+        lines(lowess(latent_scores[[g]][ ,paste0("fs_", latent_iv[i])], df_latent_scores[[g]]$residuals),
+              col = "red")
+      }
+    }
+
+  } else {
+    latent_names <- colnames(latent_scores)
```

Comment:
Is this part the same as the ones in the multiple-group version? Would it be more efficient write a function for single group, and then if the input is multiple group, you can just call the the single-group function multiple times?